parcels/_interpolation.py

```diff
@@ -0,0 +1,276 @@
+    return (1 - ctx.zeta) * f0 + ctx.zeta * f1
+
+
+@register_3d_interpolator("linear_invdist_land_tracer")
+def _linear_invdist_land_tracer_3d(ctx: InterpolationContext3D) -> float:
+    land = np.isclose(ctx.data[ctx.ti, ctx.zi : ctx.zi + 2, ctx.yi : ctx.yi + 2, ctx.xi : ctx.xi + 2], 0.0)
+    nb_land = np.sum(land)
+    if nb_land == 8:
+        return 0
+    elif nb_land > 0:
+        val = 0
+        w_sum = 0
+        for k in range(2):
+            for j in range(2):
+                for i in range(2):
+                    distance = pow((ctx.zeta - k), 2) + pow((ctx.eta - j), 2) + pow((ctx.xsi - i), 2)
+                    if np.isclose(distance, 0):
+                        if land[k][j][i] == 1:  # index search led us directly onto land
+                            return 0
+                        else:
+                            return ctx.data[ctx.ti, ctx.zi + k, ctx.yi + j, ctx.xi + i]
+                    elif land[k][j][i] == 0:
+                        val += ctx.data[ctx.ti, ctx.zi + k, ctx.yi + j, ctx.xi + i] / distance
+                        w_sum += 1 / distance
+        return val / w_sum
+    else:
+        data = ctx.data[ctx.ti, ctx.zi, :, :]
+        f0 = (
+            (1 - ctx.xsi) * (1 - ctx.eta) * data[ctx.yi, ctx.xi]
+            + ctx.xsi * (1 - ctx.eta) * data[ctx.yi, ctx.xi + 1]
+            + ctx.xsi * ctx.eta * data[ctx.yi + 1, ctx.xi + 1]
+            + (1 - ctx.xsi) * ctx.eta * data[ctx.yi + 1, ctx.xi]
+        )
+        data = ctx.data[ctx.ti, ctx.zi + 1, :, :]
+        f1 = (
+            (1 - ctx.xsi) * (1 - ctx.eta) * data[ctx.yi, ctx.xi]
+            + ctx.xsi * (1 - ctx.eta) * data[ctx.yi, ctx.xi + 1]
+            + ctx.xsi * ctx.eta * data[ctx.yi + 1, ctx.xi + 1]
+            + (1 - ctx.xsi) * ctx.eta * data[ctx.yi + 1, ctx.xi]
+        )
+        return (1 - ctx.zeta) * f0 + ctx.zeta * f1
+
+
+@register_3d_interpolator("linear")
+@register_3d_interpolator("bgrid_velocity")
+@register_3d_interpolator("bgrid_w_velocity")
+@register_3d_interpolator("partialslip")
+@register_3d_interpolator("freeslip")
+def _linear_3d_old(ctx: InterpolationContext3D) -> float:
+    zeta = ctx.zeta
+    eta = ctx.eta
+    xsi = ctx.xsi
+    ti = ctx.ti
+    zi = ctx.zi
+    xi = ctx.xi
+    yi = ctx.yi
+    if ctx.interp_method == "bgrid_velocity":
+        if ctx.gridindexingtype == "mom5":
+            zeta = 1.0
+        else:
+            zeta = 0.0
+    elif ctx.interp_method == "bgrid_w_velocity":
+        eta = 1.0
+        xsi = 1.0
+    data = ctx.data[ti, zi, :, :]
+    f0 = (
+        (1 - xsi) * (1 - eta) * data[yi, xi]
+        + xsi * (1 - eta) * data[yi, xi + 1]
+        + xsi * eta * data[yi + 1, xi + 1]
+        + (1 - xsi) * eta * data[yi + 1, xi]
+    )
+    if ctx.gridindexingtype == "pop" and zi >= ctx.grid.zdim - 2:
+        # Since POP is indexed at cell top, allow linear interpolation of W to zero in lowest cell
+        return (1 - zeta) * f0
```

Comment:
I'm looking at this code, and would like to remove the dependency on the `grid` class. Do you know whether chunking also happens vertically @erikvansebille? wondering whether `data.shape[1]` would be the equivalent of `grid.zdim`.

Otherwise I might need to have a deeper think on this

Reply:
I can't see any handling of chunks in the scipy version of the code

Reply:
> I can't see any handling of chunks in the scipy version of the code, at least not in field.py or these new files, so perhaps `data` is full?

Yep, indeed there is no explicit handling of the chunks in script; to be honest I'm not sure whether that's because the data is converted to full at some point, or whether that magically happens somewhere by dask/numpy under the hood...

Reply:
> I'm looking at this code, and would like to remove the dependency on the `grid` class. Do you know whether chunking also happens vertically @erikvansebille? wondering whether `data.shape[1]` would be the equivalent of `grid.zdim`. Note that `data` is just `Field.data`
> 

Yep, I think you're right that `data.shape[1]` will always be zdim; although it could be that we have some cases were `len(data)=3`, and `data[0]` is then _either_ depth or time. In that case, it wouldn't work... 

This was the implementation in earlier Parcels versions, but I can't remember if we harmonised it for `len(data)` to always be `4`. That would certainly make the logic easier!

Anyways, I'd give it a try, if I were you. The passing/failing of unit tests will give you an idea whether it works or not...